### DISCUSSION TITLE
Add Last Awake State for Battery Devices to UI

### DISCUSSION THREAD

**Discussion by trickiegt** [zwave-js/zwave-js-ui#2962] [Ideas]
Add Last Awake State for Battery Devices to UI

I imagine this would be relatively easy to implement and could be quite helpful / informative.
My thoughts are... when you hover over the zzz Icon, currently is says "Asleep."  I think it would be fantastic if instead it would show the last time the node was awake.  Similar to 'last active' but this would be beneficial if you wanted to change a nodes configuration or for healing purposes without going to manually wake the device or digging through logs looking for the last time it was awake.

Bonus thought but fairly irrelevant -- You could even calculate the nodes next awake time! 

Thanks for reading and all the hard work and effort that has gone into this project!  I absolutely love it and salute everyone involved!!

-Chris

**Comment by robertsLando** [comment]
@trickiegt The already present last active column it's the same, I update it each time controller receives a packet from that device (so when device was awake)

**Comment by trickiegt** [comment]
Respectfully, last active and last awake are not the same.  While yes, when a node wakes, the node will send data to the controller notifying the controller that it is awake.  When that happens, the node is active and updates the last active state.  However, for example, when a motion sensor detects motion... The sensor will relay information back to the controller telling it there was motion detected.  When the controller receives this information, it will update the nodes last active timestamp.  However, just because the node communicated with the controller does not mean the node was awake.

When nodes are awake, they can receive information from the controller.
When they are asleep, they can only send information to the controller.

I hope I'm not being rude, I truly do appreciate and respect all of your hard work and effort.  It was just a thought and I can use Home Assistant to get the last wake time without digging through log files.  So no worries, thanks again for this wonderful UI.

-Chris

**Reply by robertsLando** [reply]
Sorry I misunderstood the question, BTW I dunno where HA gets this information but it's not difficult to add anyway, I would like to know if they get them directly from zwave-js driver or if they simply store it on their db when the event happens.

cc @AlCalzone @kpine any clue about this?

**Reply by kpine** [reply]
HA has a "node status" sensor. It listens for [sleep/wake](https://zwave-js.github.io/node-zwave-js/#/api/node?id=quotwake-upquot-quotsleepquot) and dead/alive node events. The state of the sensor is then updated to reflect the [node status property](https://zwave-js.github.io/node-zwave-js/#/api/node?id=status) at the time of the event.

Most entities that have state changes are recorded in the HA state database. I don't know if users have access to the history, however it's simple to create your own sensor to capture the last awake time.

The HA state information is not going to be 100% reliable in this case. If HA is offline while Z-Wave JS is online, a wake event will never be received, and if HA restarts and the state has changed, the timestamp would reflect the time HA received it, not when it actually happened. This is not a problem for ZUI since it is only online when the driver is. In either case, there is nothing that can be done if the controller is offline and the device wakes though, or if an error in transmission got lost or corrupted, so there's always a chance for it to be missed.

So if you're going to base some decisions on "last awake", then you need to tolerate these potential issues, hence [why](https://github.com/zwave-js/node-zwave-js/issues/4207#issuecomment-1034196905) the driver currently doesn't handle it natively.

Also, it's fairly easy to manually find a node's last wake up notification time, no need to dig through logs. ZUI already logs all events.

![image](https://user-images.githubusercontent.com/791794/218794732-6e7abd09-3af9-4806-8aca-a8cd962c28fd.png)

**Reply by AlCalzone** [reply]
@robertsLando Similar to your last  active column, you just need to have a value that's updated when the device status changes to awake.

**Reply by trickiegt** [reply]
Thanks for the intel guys -- in HA, you can easily see the history in the devices log list.  I never looked at the events in node list, nor did I realize it is searchable... Very nice to know for future reference.  I was just thinking it would be nice to see a quick detail on the main overview page that also took into account timezone data so I can avoid doing math =)  Pretty much the exact same concept as "Last Active"  But I would display the time when hovering over the zzz icon instead of it saying "Asleep"

**Reply by robertsLando** [reply]
Ok no problem I can do it. Only problem I see with this is that I will always start updating that field from when zui starts so if you restart zui you need your nodes to wake up in order to see a timestamp there (will not be persisted), like it happens for last active

**Reply by trickiegt** [reply]
That sounds fantastic, thanks for hearing me out and having interest in implementing it!

**Reply by robertsLando** [reply]
Always open to implement new features if them could be useful :)

**Comment by LevinXX** [comment]
Hi

Just wondering if the "last wake up" value has been implemented? Can't find it in the UI.

It will be very helpful because I have many battery devices and I was able to get this value with the openHAB native Z-Wave binding.
Now I moved to the zwave-js-ui where I can only see "last active"

**Reply by robertsLando** [reply]
merged to master now, I may have miss this
